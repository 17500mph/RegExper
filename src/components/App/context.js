import React from 'react';
import PropTypes from 'prop-types';
import URLSearchParams from '@ungap/url-search-params';

const AppContext = React.createContext();

const toUrl = params => new URLSearchParams(params).toString();

const readURLHash = () => {
  const query = document.location.hash.slice(1);
  const params = new URLSearchParams(query);

  if (params.get('syntax')) {
    return {
      syntax: params.get('syntax'),
      expr: params.get('expr')
    };
  } else {
    // Assuming old-style URL
    return {
      syntax: 'js',
      expr: query || ''
    };
  }
};

const createSvgLink = async ({ svg }) => {
  try {
    const type = 'image/svg+xml';
    const blob = new Blob([svg], { type });

    return {
      url: URL.createObjectURL(blob),
      label: 'Download SVG',
      filename: 'image.svg',
      type
    };
  }
  catch (e) {
    console.error(e); // eslint-disable-line no-console
  }
};

const createPngLink = async ({ svg, width, height }) => {
  try {
    const type = 'image/png';
    const canvas = document.createElement('canvas');
    const context = canvas.getContext('2d');
    const loader = new Image();

    loader.width = canvas.width = Number(width) * 2;
    loader.height = canvas.height = Number(height) * 2;

    await new Promise(resolve => {
      loader.onload = resolve;
      loader.src = 'data:image/svg+xml,' + encodeURIComponent(svg);
    });

    context.drawImage(loader, 0, 0, loader.width, loader.height);

    const blob = await new Promise(resolve => canvas.toBlob(resolve, type));

    return {
      url: URL.createObjectURL(blob),
      label: 'Download PNG',
      filename: 'image.png',
      type
    };
  }
  catch (e) {
    console.error(e); // eslint-disable-line no-console
  }
};

class AppContextProvider extends React.PureComponent {
  state = {}

  mutations = {
    svgData: async ({ svg, width, height }) => {
      if (svg !== this.state.svg) {
        this.setState({
          svg,
          svgLink: await createSvgLink({ svg }),
          pngLink: await createPngLink({ svg, width, height })
        });
      }
    },

    renderExpr: ({ syntax, expr }) => {
      if (expr) {
        document.location.hash = toUrl({ syntax, expr });
      }
    }
  }

  componentDidMount() {
    // Gatsby doesn't have document.location, so readURLHash can't be called
    // until here
    this.setState(readURLHash());

    window.addEventListener('hashchange', this.handleHashChange);
    this.handleHashChange();
  }

  componentWillUnmount() {
    window.removeEventListener('hashchange', this.handleHashChange);
  }

  handleHashChange = () => {
    const { expr, syntax } = readURLHash();

    if (!expr) {
      return;
    }

    this.setState({
      syntax,
      expr,
      permalinkUrl: document.location.toString()
    });

    if (this.props.onExpressionChange) {
      this.props.onExpressionChange({ syntax, expr });
    }
  }

  render() {
    const { children } = this.props;
    const context = {
      ...this.state,
      ...this.mutations
    };

    return <AppContext.Provider value={ context }>
      { children }
    </AppContext.Provider>;
  }
}

AppContextProvider.propTypes = {
  onExpressionChange: PropTypes.func,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]).isRequired
};

export { AppContextProvider };
export default AppContext;
