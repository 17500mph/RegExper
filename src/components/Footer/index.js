import React from 'react';
import PropTypes from 'prop-types';
import { StaticQuery, graphql } from 'gatsby';
import { withNamespaces, Trans } from 'react-i18next';

import style from './style.module.css';

const query = graphql`
  query FooterQuery {
    site {
      siteMetadata {
        buildId
      }
    }
  }
`;

export const Footer = ({ t, buildId }) => (
  <footer className={ style.footer }>
    <ul className={ style.list }>
      <li>
        <Trans>Created by <a
          href="mailto:jeff.avallone@gmail.com">Jeff Avallone</a></Trans>
      </li>
      <li>
        <Trans>Generated images licensed: <a
          href="http://creativecommons.org/licenses/by/3.0/"
          rel="license external noopener noreferrer"
          target="_blank">
          <img src="https://licensebuttons.net/l/by/3.0/80x15.png"
            alt={ t('Creative Commons CC-BY-3.0 License') } />
        </a></Trans>
      </li>
    </ul>
    <div className={ style.buildId }>
      { buildId }
    </div>
  </footer>
);

Footer.propTypes = {
  t: PropTypes.func.isRequired,
  buildId: PropTypes.string.isRequired
};

export default withNamespaces()(props => (
  <StaticQuery query={ query } render={ ({ site: { siteMetadata } }) => (
    <Footer { ...props } { ...siteMetadata } />
  ) } />
));
