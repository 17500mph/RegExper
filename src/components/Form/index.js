import React from 'react';
import PropTypes from 'prop-types';
import { translate, Trans } from 'react-i18next';

import DownloadIcon from 'feather-icons/dist/icons/download.svg';
import LinkIcon from 'feather-icons/dist/icons/link.svg';
import ExpandIcon from 'feather-icons/dist/icons/chevrons-down.svg';

import style from './style.css';

class Form extends React.PureComponent {
  state = {
    expr: this.props.expr,
    syntax: this.props.syntax || Object.keys(this.props.syntaxes)[0]
  }

  handleSubmit = event => {
    event.preventDefault();
    this.props.onSubmit.call(this, {
      expr: this.state.expr,
      syntax: this.state.syntax
    });
  }

  handleKeyPress = event => {
    if (event.charCode === 13 && event.shiftKey) {
      this.handleSubmit(event);
    }
  }

  handleChange = event => this.setState({ [event.target.name]: event.target.value })

  permalinkAction() {
    const { permalinkUrl } = this.props;

    if (!permalinkUrl) {
      return;
    }

    return <li>
      <a href={ permalinkUrl }><LinkIcon/><Trans>Permalink</Trans></a>
    </li>;
  }

  downloadActions() {
    const { downloadUrls } = this.props;

    if (!downloadUrls) {
      return;
    }

    return downloadUrls.map(({ url, filename, type, label }, i) => <li key={ i }>
      <a href={ url } download={ filename } type={ type }>
        <DownloadIcon/><Trans>{ label }</Trans>
      </a>
    </li>);
  }

  render() {
    const { syntaxes, t } = this.props;
    const { expr, syntax } = this.state;

    return <div className={ style.form }>
      <form onSubmit={ this.handleSubmit }>
        <textarea
          name="expr"
          value={ expr }
          onKeyPress={ this.handleKeyPress }
          onChange={ this.handleChange }
          autoFocus
          placeholder={ t('Enter regular expression to display') }></textarea>
        <button type="submit"><Trans>Display</Trans></button>
        <div className={ style.select }>
          <select
            name="syntax"
            value={ syntax }
            onChange={ this.handleChange }>
            { Object.keys(syntaxes).map(id => (
              <option value={ id } key={ id }>{ syntaxes[id] }</option>
            )) }
          </select>
          <ExpandIcon/>
        </div>
        <ul className={ style.actions }>
          { this.downloadActions() }
          { this.permalinkAction() }
        </ul>
      </form>

    </div>;
  }
}

Form.propTypes = {
  expr: PropTypes.string,
  syntax: PropTypes.string,
  syntaxes: PropTypes.object,
  onSubmit: PropTypes.func,
  permalinkUrl: PropTypes.string,
  downloadUrls: PropTypes.array,
  t: PropTypes.func
};

export default translate()(Form);
export { Form };
