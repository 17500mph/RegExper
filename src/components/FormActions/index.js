import React from 'react';
import PropTypes from 'prop-types';
import { withNamespaces, Trans } from 'react-i18next';

import DownloadIcon from 'react-feather/dist/icons/download';
import LinkIcon from 'react-feather/dist/icons/link';

import style from './style.module.css';

import { createPngLink, createSvgLink } from './links';

class FormActions extends React.PureComponent {
  static propTypes = {
    permalinkUrl: PropTypes.string,
    imageDetails: PropTypes.shape({
      svg: PropTypes.string,
      width: PropTypes.number,
      height: PropTypes.number
    }),
    t: PropTypes.func.isRequired
  }

  state = {
    svgLink: null,
    pngLink: null
  }

  componentDidMount() {
    const { imageDetails } = this.props;

    if (imageDetails && imageDetails.svg) {
      this.generateDownloadLinks();
    }
  }

  componentDidUpdate(prevProps) {
    const { imageDetails } = this.props;
    const { imageDetails: prevImageDetails } = prevProps;

    if (!imageDetails) {
      this.setState({ svgLink: null, pngLink: null });
      return;
    }

    if (!prevImageDetails) {
      this.generateDownloadLinks();
      return;
    }

    if (imageDetails.svg !== prevImageDetails.svg
      || imageDetails.width !== prevImageDetails.width
      || imageDetails.height !== prevImageDetails.height) {
      this.generateDownloadLinks();
      return;
    }
  }

  async generateDownloadLinks() {
    const { imageDetails: { svg, width, height } } = this.props;

    this.setState({
      svgLink: await createSvgLink({ svg }),
      pngLink: await createPngLink({ svg, width, height })
    });
  }

  downloadLink({ url, filename, type, label }) {
    const { t } = this.props;

    return <li>
      <a href={ url } download={ filename } type={ type }>
        <DownloadIcon />{ t(label) }
      </a>
    </li>;
  }

  render() {
    const {
      permalinkUrl
    } = this.props;
    const {
      svgLink,
      pngLink
    } = this.state;

    return <ul className={ style.actions }>
      { pngLink && this.downloadLink(pngLink) }
      { svgLink && this.downloadLink(svgLink) }
      { permalinkUrl && <li>
        <a href={ permalinkUrl }><LinkIcon /><Trans>Permalink</Trans></a>
      </li> }
    </ul>;
  }
}

export { FormActions };
export default withNamespaces()(FormActions);
