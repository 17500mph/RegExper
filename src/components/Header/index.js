import React from 'react';
import { translate, Trans } from 'react-i18next';

import style from './style.css';
import GitlabIcon from 'feather-icons/dist/icons/gitlab.svg';

import LocaleSwitcher from 'components/LocaleSwitcher';

const Header = () => (
  <header className={ style.header } data-banner={ process.env.BANNER }>
    <h1>
      <a href="/">Regexper</a>
    </h1>

    <ul>
      <li><a href="https://gitlab.com/javallone/regexper-static" rel="external noopener noreferrer" target="_blank">
        <GitlabIcon/><Trans>Source on GitLab</Trans>
      </a></li>
      <li>
        <a href="/privacy.html"><Trans>Privacy Policy</Trans></a>
      </li>
      <li><LocaleSwitcher /></li>
    </ul>
  </header>
);

export default translate()(Header);
export { Header };
