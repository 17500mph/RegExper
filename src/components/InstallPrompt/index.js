import React from 'react';
import { withTranslation, Trans } from 'react-i18next';

class InstallPrompt extends React.PureComponent {
  state = {
    installPrompt: null
  }

  componentDidMount() {
    window.addEventListener('beforeinstallprompt', this.handleInstallPrompt);
  }

  componentWillUnmount() {
    window.removeEventListener('beforeinstallprompt', this.handleInstallPrompt);
  }

  handleInstallPrompt = event => {
    this.setState({
      installPrompt: event
    });
  }

  handleInstall = async event => {
    event.preventDefault();

    const { installPrompt } = this.state;

    try {
      installPrompt.prompt();
      await installPrompt.userChoice;
    }
    catch {
      // User cancelled install
    }

    this.setState({ installPrompt: null });
  }

  render() {
    const { installPrompt } = this.state;

    if (!installPrompt) {
      return null;
    }

    return <a href="#install" onClick={ this.handleInstall }>
      <Trans>Add to Home Screen</Trans>
    </a>;
  }
}

export { InstallPrompt };
export default withTranslation()(InstallPrompt);
