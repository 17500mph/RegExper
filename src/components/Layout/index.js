import React from 'react';
import PropTypes from 'prop-types';
import { Helmet } from 'react-helmet';

import Header from 'components/Header';

const Layout = ({ title, children }) => <React.Fragment>
  <Helmet>
    <title>{ title ? `Regexper - ${ title }` : 'Regexper' }</title>
  </Helmet>
  <Header />
  { children }
</React.Fragment>;

Layout.propTypes = {
  title: PropTypes.string,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]).isRequired
};

export default Layout;
