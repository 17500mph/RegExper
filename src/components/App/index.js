import React from 'react';
import PropTypes from 'prop-types';
import { translate } from 'react-i18next';
import URLSearchParams from 'url-search-params';
import Raven from 'raven-js';

import LoaderIcon from 'feather-icons/dist/icons/loader.svg';

import style from './style.css';

import Form from 'components/Form';
import Message from 'components/Message';
import InstallPrompt from 'components/InstallPrompt';
import { demoImage } from 'devel';

const syntaxes = {
  js: 'JavaScript',
  pcre: 'PCRE'
};

const toUrl = params => new URLSearchParams(params).toString();

class App extends React.PureComponent {
  state = {}

  image = React.createRef()

  componentDidMount() {
    window.addEventListener('hashchange', this.handleHashChange);
    window.addEventListener('beforeinstallprompt', this.handleInstallPrompt);
    this.handleHashChange();
  }

  componentWillUnmount() {
    window.removeEventListener('hashchange', this.handleHashChange);
    window.removeEventListener('beforeinstallprompt', this.handleInstallPrompt);
  }

  async setSvgUrl() {
    try {
      const type = 'image/svg+xml';
      const blob = await this.image.current.svgUrl(type);

      this.setState({
        svgUrl: {
          url: URL.createObjectURL(blob),
          label: 'Download SVG',
          filename: 'image.svg',
          type
        }
      });
    }
    catch (e) {
      console.error(e); // eslint-disable-line no-console
    }
  }

  async setPngUrl() {
    try {
      const type = 'image/png';
      const blob = await this.image.current.pngUrl(type);

      this.setState({
        pngUrl: {
          url: URL.createObjectURL(blob),
          label: 'Download PNG',
          filename: 'image.png',
          type
        }
      });
    }
    catch (e) {
      console.error(e); // eslint-disable-line no-console
    }
  }

  async loadSVGComponent() {
    if (this.state.SVG) {
      return;
    }

    this.setState({
      loading: true,
      loadingFailed: false
    });

    try {
      const SVG = await import(/* webpackChunkName: "render" */ 'components/SVG');

      this.setState({
        SVG: SVG.default,
        loading: false
      });
    }
    catch (e) {
      Raven.captureException(e);
      this.setState({
        loading: false,
        loadingFailed: e
      });
      throw e;
    }
  }

  handleInstallPrompt = event => {
    event.preventDefault();

    this.setState({
      installPrompt: event
    });
  }

  handleSubmit = ({expr, syntax}) => {
    if (expr) {
      document.location.hash = toUrl({ syntax, expr });
    }
  }

  handleHashChange = async () => {
    const query = document.location.hash.slice(1);
    const params = new URLSearchParams(query);
    const { expr, syntax } = (() => {
      if (params.get('syntax')) {
        return {
          syntax: params.get('syntax'),
          expr: params.get('expr')
        };
      } else {
        // Assuming old-style URL
        return {
          syntax: 'js',
          expr: query
        };
      }
    })();

    if (!expr) {
      return;
    }

    try {
      await this.loadSVGComponent();
      console.log(syntax, expr); // eslint-disable-line no-console
      this.setState({
        image: demoImage,
        permalinkUrl: document.location.toString(),
        syntax,
        expr
      }, async () => {
        await this.image.current.doReflow();
        this.setSvgUrl();
        this.setPngUrl();
      });
    }
    catch (e) {
      console.error(e); // eslint-disable-line no-console
    }
  }

  handleRetry = async event => {
    event.preventDefault();
    this.handleHashChange();
  }

  handleInstallReject = () => {
    this.setState({ installPrompt: null });
  }

  handleInstallAccept = async () => {
    const { installPrompt } = this.state;

    this.setState({ installPrompt: null });
    installPrompt.prompt();
  }

  render() {
    const {
      SVG,
      loading,
      loadingFailed,
      svgUrl,
      pngUrl,
      permalinkUrl,
      syntax,
      expr,
      image,
      installPrompt
    } = this.state;
    const downloadUrls = [
      svgUrl,
      pngUrl
    ].filter(Boolean);

    return <React.Fragment>
      <Form
        key={ toUrl({ expr, syntax }) }
        syntaxes={ syntaxes }
        downloadUrls={ downloadUrls }
        permalinkUrl={ permalinkUrl }
        syntax={ syntax }
        expr={ expr }
        onSubmit={ this.handleSubmit }/>
      {
        loading && <div className={ style.loader }>
          <LoaderIcon />
          <div className={ style.message }>Loading...</div>
        </div>
      }
      {
        loadingFailed && <Message type="error" heading="Render Failure">
          An error occurred while rendering the regular expression. <a href="#retry" onClick={ this.handleRetry }>Retry</a>
        </Message>
      }
      {
        image && <div className={ style.render }>
          <SVG data={ image } ref={ this.image }/>
        </div>
      }
      {
        installPrompt && <InstallPrompt onAccept={ this.handleInstallAccept } onReject={ this.handleInstallReject } />
      }
    </React.Fragment>;
  }
}

App.propTypes = {
  t: PropTypes.func
};

export default translate()(App);
export { App };
