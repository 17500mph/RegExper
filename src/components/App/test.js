import React from 'react';
import { shallow } from 'enzyme';

import { mockT } from 'i18n';
import { App } from 'components/App';

jest.mock('syntax/js', () => ({
  parse: expr => `PARSED(${ expr })`,
  layout: parsed => `LAYOUT(${ parsed })`,
  Render: () => ''
}));

const syntaxList = [
  { id: 'testJS', label: 'Testing JS' },
  { id: 'other', label: 'Other' }
];
const commonProps = { syntaxList, t: mockT };

describe('App', () => {
  test('rendering', () => {
    const component = shallow(
      <App expr="" syntax="js" { ...commonProps } />
    );
    expect(component).toMatchSnapshot();
  });

  test('rendering an expression', async () => {
    const component = shallow(
      <App expr="" syntax="js" { ...commonProps } />
    );
    expect(component).toMatchSnapshot();

    component.setProps({
      expr: 'test expression'
    });

    expect(component).toMatchSnapshot();

    // Give a beat for module to load
    await new Promise(resolve => setTimeout(resolve));

    expect(component).toMatchSnapshot();
  });

  test('rendering with an invalid syntax', async () => {
    jest.spyOn(console, 'error').mockImplementation(() => {});

    const component = shallow(
      <App expr="" syntax="invalid" { ...commonProps } />
    );
    expect(component).toMatchSnapshot();

    component.setProps({
      expr: 'test expression'
    });

    expect(component).toMatchSnapshot();

    // Give a beat for module to load
    await new Promise(resolve => setTimeout(resolve));

    expect(component).toMatchSnapshot();
  });

  test('removing rendered expression', async () => {
    const component = shallow(
      <App expr="test expression" syntax="js" { ...commonProps } />
    );

    // Give a beat for module to load
    await new Promise(resolve => setTimeout(resolve));

    expect(component).toMatchSnapshot();

    component.setProps({
      expr: ''
    });

    expect(component).toMatchSnapshot();
  });

  test('rendering image details', async () => {
    const component = shallow(
      <App expr="test expression" syntax="js" { ...commonProps } />
    );

    // Give a beat for module to load
    await new Promise(resolve => setTimeout(resolve));

    expect(component).toMatchSnapshot();

    component.instance().handleSvg({
      svg: 'test svg content'
    });

    expect(component).toMatchSnapshot();
  });

  test('retrying expression rendering', () => {
    jest.spyOn(console, 'error').mockImplementation(() => {});

    const component = shallow(
      <App expr="test expression" syntax="invalid" { ...commonProps } />
    );

    const instance = component.instance();
    const event = { preventDefault: jest.fn() };

    jest.spyOn(instance, 'handleRender');

    instance.handleRetry(event);

    expect(event.preventDefault).toHaveBeenCalled();
    expect(instance.handleRender).toHaveBeenCalled();
  });

  test('submitting an expression to render', () => {
    const component = shallow(
      <App expr="" syntax="js" { ...commonProps } />
    );

    const instance = component.instance();

    instance.handleSubmit({ syntax: 'test', expr: '' });

    expect(document.location.hash).toEqual('');

    instance.handleSubmit({ syntax: 'test', expr: 'test expression' });

    expect(document.location.hash).toEqual('#syntax=test&expr=test+expression');
  });
});
