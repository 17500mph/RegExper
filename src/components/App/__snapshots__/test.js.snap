// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App rendering 1`] = `
<React.Fragment>
  <Translate(Form)
    downloadUrls={Array []}
    key="expr=undefined&syntax=undefined"
    onSubmit={[Function]}
    syntaxes={
      Object {
        "js": "JavaScript",
        "pcre": "PCRE",
      }
    }
  />
</React.Fragment>
`;
